src/Tribe/Rewrite.php

```diff
@@ -295,4 +302,232 @@ protected function replace_array_key( &$array, $search, $replace ) {
+		) {
+			$wp_canonical = redirect_canonical( $canonical_url, false );
+			if ( empty( $wp_canonical ) ) {
+				$wp_canonical = $canonical_url;
+			}
+
+			$this->canonical_url_cache[ $url ] = $wp_canonical;
+
+			return $wp_canonical;
+		}
+
+		$bases = (array) $this->get_bases();
+		ksort( $bases );
+
+		$localized_matchers = $this->get_localized_matchers();
+		$dynamic_matchers   = $this->get_dynamic_matchers( $query_vars );
+
+		// Try to match only on the query vars we're actually handling.
+		$matched_vars   = array_intersect_key( $query_vars, array_combine( $handled_query_vars, $handled_query_vars ) );
+		$unmatched_vars = array_diff_key( $query_vars, array_combine( $handled_query_vars, $handled_query_vars ) );
+
+		if ( empty( $matched_vars ) ) {
+			// The URL does contain query vars, but none we handle.
+			$wp_canonical = trailingslashit( redirect_canonical( $url, false ) );
+			$this->canonical_url_cache[ $url ] = $wp_canonical;
+
+			return $wp_canonical;
+		}
+
+		foreach ( $our_rules as $link_template => $index_path ) {
+			wp_parse_str( (string) parse_url( $index_path, PHP_URL_QUERY ), $link_vars );
+			ksort( $link_vars );
+
+			if ( array_keys( $link_vars ) !== array_keys( $matched_vars ) ) {
+				continue;
+			}
+
+			if ( ! (
+				Arr::get( $matched_vars, 'post_type', '' ) === Arr::get( $link_vars, 'post_type', '' )
+				&& Arr::get( $matched_vars, 'eventDisplay', '' ) === Arr::get( $link_vars, 'eventDisplay', '' )
+			) ) {
+				continue;
+			}
+
+			$replace = array_map( static function ( $index ) use ( $matched_vars ) {
+				return isset( $matched_vars[ $index ] )
+					? str_replace( 'tribe_', '', $matched_vars[ $index ] )
+					: '';
+			}, $localized_matchers );
+			// Include dynamic matchers now.
+			$replace = array_merge( $dynamic_matchers, $replace );
+
+			$replaced = str_replace( array_keys( $replace ), $replace, $link_template );
+			// Remove trailing chars.
+			$path          = rtrim( $replaced, '?$' );
+			$canonical_url = trailingslashit( home_url( $path ) );
+
+			break;
+		}
+
+		$wp_canonical = redirect_canonical( $canonical_url, false );
+		$resolved     = empty( $wp_canonical ) ? $canonical_url : $wp_canonical;
+		if ( $canonical_url !== $resolved ) {
+			$resolved = trailingslashit( $resolved );
+		}
+
+		if ( count( $unmatched_vars ) ) {
+			$resolved = add_query_arg( $unmatched_vars, $canonical_url );
+		}
+
+		$this->canonical_url_cache[ $url ] = $resolved;
+
+		return $resolved;
+	}
+
+	/**
+	 * Returns an array of rewrite rules handled by the implementation.
+	 *
+	 * @since TBD
+	 *
+	 * @return array An array of rewrite rules handled by the implementation in the shape `[ <regex> => <path> ]`.
+	 */
+	protected function get_handled_rewrite_rules() {
+		global $wp_rewrite;
+		// While this is specific to The Events Calendar we're handling a small enough post type base to keep it here.
+		$pattern = '/post_type=tribe_(events|venue|organizer)/';
```

Comment:
Should we have a filter for post types that we support and then pass those into a pattern?